src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1076,7 +1076,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val sam = samMatchingFunction(tree, pt) // this implies tree.isInstanceOf[Function]
           if (sam.exists) {
             val samTree = adaptToSAM(sam, tree.asInstanceOf[Function], pt, mode)
-            if (samTree ne EmptyTree) return samTree
+            if (samTree ne EmptyTree)
+              return samTree.updateAttachment(SAMFunction(sam.owner))
```

Comment:
I propose that pass the SAM method (rather than just its owner.)

We need this info in the backend to pass to the `LambdaMetaFactory` bootstrap. I currently uses a [scheme](https://github.com/scala/scala/blob/c8e6050c3c190dd064642b6b77fc179f27b0495d/src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala#L1378-L1378) that is hard-coded to `FunctionN` / `JFunctionN` (look for the single `deferred` method in `decls` or instead use the `apply` method).

This doesn't work in general for SAMs:

```
scala> trait T { def foo(a: Any): Any }; trait U extends T;
defined trait T
defined trait U

scala> def foo: U = (x: Any) => x
java.lang.AssertionError: assertion failed:
  not a method-symbol: <none>
     while compiling: <console>
        during phase: jvm
     library version: version 2.12.0-20160321-153935-3ff165f
    compiler version: version 2.12.0-20160321-153935-3ff165f
  reconstructed args: -usejavacp

  last tree to typer: Function(value $anonfun)
       tree position: line 13 of <console>
            tree tpe: iw$U
              symbol: value $anonfun
   symbol definition: val $anonfun: <notype> (a TermSymbol)
      symbol package: $line4
       symbol owners: value $anonfun -> method foo -> object iw
           call site: constructor $read$$iw$$iw in package $line4

<Cannot read source file>
	at scala.tools.nsc.Global.assert(Global.scala:245)
	at scala.tools.nsc.backend.jvm.BTypesFromSymbols.methodBTypeFromSymbol(BTypesFromSymbols.scala:132)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genInvokeDynamicLambda(BCodeBodyBuilder.scala:1380)
```

IIRC, the slightly convoluted scheme for figuring out this method for `JFunctionN` was related to quirks specializion, so I'm not sure if we "dead reckon" to the right method when we create the `LambdaMetatfactoryCapable` attachment in `delamdafy` in order to unify code gen for SAM and regular functions.